### ISSUE TITLE
Issues with webapp.hs demo

### ISSUE THREAD

**Issue by mephistopheles-8** [agocorona/ghcjs-hplay#22]
I attempted a build of the webapp.hs demo.  It kinda worked! I did run into a few issues, though:

- The output js attempted to call parentNode of an HTMLElement as a function.  I just changed this in the output javascript to get the demo to work.
- A few errors were received in the console, including `listen websocket: request received for non existent closure` and `listen: Communication error`

The Fibonacci sequence and the input bits appeared to work fine, though there was some flickering in the browser during the socket errors. The counter buttons crashed the program.  

I'm using the lts-7.19 resolver, and my extra deps look like this:

    extra-deps:
    - axiom-0.4.6
    - transient-0.5.9.2
    - transient-universe-0.4.6.1

I saw that the stack.yaml in the repos were fixed to commits -- should I do the same?

**Comment by agocorona** [comment]
Hi @mephistopheles-8 

Thanks for the feedback

> The output js attempted to call parentNode of an HTMLElement as a function. I just changed this in the output javascript to get the demo to work

What do you mean by that?  did you fixed the javascript code?

**Comment by agocorona** [comment]
Please give me your OS, version of GHC and GHCJS
Maybe you used the javascript code existent in the static folder.  Thar code is not valid for that version of Axiom. you have to install the libraries for GHCJS and compile  webapp.hs with them:

install GHCJS

     cabal install  axiom --ghcjs
     ghcjs webapp.hs -o static/out

then execute webapp

or alternatively, use the docker instance:

    webapp.hs -p start/localhost/8080


That uses a docker image that compiles the program with ghcjs  and ghc and executes the program.  you should put the IP of the docker image.  In my case http://192.168.99.100:8080

**Comment by agocorona** [comment]
If you read this in the mail, I updated the previous message, please see it in github

**Comment by mephistopheles-8** [comment]
Yes, I edited the js output.  I'm using Ubuntu 16.04, with this version of ghcjs:

    compiler-check: match-exact
    compiler: ghcjs-0.2.1.9007019_ghc-8.0.1
    setup-info:
      ghcjs:
        source:
          ghcjs-0.2.1.9007019_ghc-8.0.1:
            url: http://ghcjs.tolysz.org/ghc-8.0-2017-02-05-lts-7.19-9007019.tar.gz
            sha1: d2cfc25f9cda32a25a87d9af68891b2186ee52f9

I  compiled both with `stack build` using separate stack.yaml files.   I copied the *.jsexe folder in the .stack-work directory to static/out.jsexe.  Not sure if it's any help, but the error was in the function below:

    function h$$aXW()
    {
      var a = h$r1;
      h$sp -= 4;
      var b = a.d1;
      var c = document.getElementById(b);
      var d = c;
      var e = (((d !== null) && !(!d.ownerDocument)) && (window === (d.ownerDocument.defaultView || d.ownerDocument.
      parentWindow)));
      if(!(!e))
      {
        var f = d.parentNode();
        h$pp24(f, h$$aXX);
        return h$e(h$ghcjszmperchzm0zi3zi3zmBXUfKl1Sw7xHXcdoD2uaflZCInternalziPerchzispan1);
      }
      else
      {
        h$l2(b, h$axiomzm0zi4zi6zmKWujBaWG9i7D2FfimKIakqZCGHCJSziHPlayziViewziwithElem2);
        return h$ap_1_1_fast();
      };
    };

Changing  `var f = d.parentNode();` to `var f = d.parentNode` made the app semi-functional. 

This was with a fresh project with deps retrieved via stack.

**Comment by agocorona** [comment]
Did you compile the program with ghcjs?

**Comment by mephistopheles-8** [comment]
Yes -- I attempted the build a second time to make sure.

I did the following:
1. Created a new project with `stack new axon simple && cd axon`
2. Added axiom, transient, and transient-universe to the cabal build-depends section.
3. Created a `static` directory in the project root
4. Copied webapp.hs to src/Main.hs
5. Created a `stack.yaml` and `stack-ghcjs.yaml` file, both with resolver lts-7.19
6. Added the version of ghcjs listed above to `stack-ghcjs.yaml`
7. Ran `stack solver --update-config` and `stack --stack-yaml=stack-ghcjs.yaml solver --update-config`
8. Ran `stack build` and `stack --stack-yaml=stack-ghcjs.yaml build`
9. Copied .stack-work/install/{path}/axon.jsexe to static/out.jsexe

Running `stack exec -- axon -p start/localhost/3000` started the app, but the page didn't load until I made the edit to the JS.

**Comment by agocorona** [comment]
It sound like an old version of ghcjs. could you execute ghcjs --version?

**Comment by mephistopheles-8** [comment]
`The Glorious Glasgow Haskell Compilation System for JavaScript, version 0.2.1.9007019 (GHC 8.0.1)`

It's the Tolysz from 2017-02-05.

**Comment by mephistopheles-8** [comment]
I see this:

https://github.com/geraldus/ghcjs-perch/commit/a902505b7aa303381907ed506dbe762802b274dc

Looks like it may have been an out-of-date version of ghcjs-perch....I will test this soon to verify.

**Comment by mephistopheles-8** [comment]
I suppose the hackage version of ghcjs-perch is not up to date.  Adding the git commit to my stack.yaml packages fixed the parentNode issue...

The second issue persists.  I still get this pretty frequently:

    listen websocket:request received for non existent closure: 204005
    CallStack (from HasCallStack):
      error, called at src/Transient/Move/Internals.hs:1230:46 in transient-universe-0.4.6.1-K8HyWOR1pM84V4YHRinFMX:Transient.Move.Internals

I have tried two configurations:

- hackage version of everything, except for the git version of ghcjs-perch: everything works, but I get the error above.
- Latest git versions of ghcjs-perch, transient, transient-universe, and axiom: the counter buttons break the demo, and I get the error above.

**Comment by agocorona** [comment]
Are you using a mac?

**Comment by agocorona** [comment]
That message is typical when the compiled program with ghc or ghcjs is not up to date
